widgets/heatmap/heatmap.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package heatmap contains a widget that displays heat maps.
+package heatmap
+
+import (
+	"errors"
+	"image"
+	"sync"
+
+	"github.com/mum4k/termdash/cell"
+	"github.com/mum4k/termdash/private/canvas"
+	"github.com/mum4k/termdash/terminal/terminalapi"
+	"github.com/mum4k/termdash/widgetapi"
+	"github.com/mum4k/termdash/widgets/heatmap/internal/axes"
+)
+
+// HeatMap draws heat map charts.
+//
+// Heatmap consists of several cells. Each cell represents a value.
+// The larger the value, the darker the color of the cell (from white to black).
+//
+// The two dimensions of the values (cells) array are determined by the length of
+// the xLabels and yLabels arrays respectively.
+//
+// HeatMap does not support mouse based zoom.
+//
+// Implements widgetapi.Widget. This object is thread-safe.
+type HeatMap struct {
+	// values are the values in the heat map.
+	values [][]float64
+
+	// xLabels are the labels on the X axis in an increasing order.
+	xLabels []string
+	// yLabels are the labels on the Y axis in an increasing order.
+	yLabels []string
+
+	// minValue and maxValue are the Min and Max values in the values,
+	// which will be used to calculate the color of each cell.
+	minValue, maxValue float64
+
+	// lastWidth is the width of the canvas as of the last time when Draw was called.
+	lastWidth int
+
+	// opts are the provided options.
+	opts *options
+
+	// mu protects the HeatMap widget.
+	mu sync.RWMutex
+}
+
+// New returns a new HeatMap widget.
+func New(opts ...Option) (*HeatMap, error) {
+	return nil, errors.New("not implemented")
+}
+
+// Values sets the values to be displayed by the HeatMap.
+//
+// Each value in values has a xLabel and a yLabel, which means
+// len(yLabels) == len(values) and len(xLabels) == len(values[i]).
+// But labels could be empty strings.
+// When no labels are provided, labels will be "0", "1", "2"...
+//
+// Each call to Values overwrites any previously provided values.
+// Provided options override values set when New() was called.
+func (hp *HeatMap) Values(xLabels []string, yLabels []string, values [][]float64, opts ...Option) error {
+	return errors.New("not implemented")
+}
+
+// ClearXLabels clear the X labels.
+func (hp *HeatMap) ClearXLabels() {
```

Comment:
Similarly to why we wanted a single function to update the the labels and values (atomicity) we will likely need a single one to clear them. At this point I am wondering if we even need a dedicated clear function. The caller can just call `Values()` again with xLabels == yLabels == nil.

What do you think?

Reply:
>  Also looks like the build is failing due to a golint error, could you fix that before we move forward?

Done

Do you mean we need add a function, like this:
```golang
func (hp *HeatMap) ClearLabels() {
    hp.xLabels = nil
    hp.yLabels = nil
}
```

I think it's ok, better than nothing, because it will not affect the implementation of HeatMap.

Reply:
Yes, we can either add a function like that, or we can just use `Values()`. Another call to values with both the `xLabels` and the `yLabels` slices empty should achieve the same result (clear labels, replace values).

Up to you really.

Reply:
I think just use `Values()` is ok.